src/thor/bidirectional_astar.cc

```diff
@@ -1016,6 +1040,9 @@ std::vector<std::vector<PathInfo>> BidirectionalAStar::FormPath(GraphReader& gra
       const BDEdgeLabel& edgelabel = edgelabels_forward_[edgelabel_index];
 
       const auto* edge = graphreader.directededge(edgelabel.edgeid(), tile);
+      if (edge == nullptr) {
```

Comment:
this just seems crazy! for this to happen it means the algorithm found the edge (meaning it had the tile at some point) then the tile disappeared but couldnt be gotten again. should we really support such  behavior!? i understand the cache could be cleared at any time by another thread but in that case shouldnt `GetGraphTile` block until it can go get the damned tile again? why would you write a tile cache implementation that has such behavior? i suppose if its network based the connection can be intermittent but still this is crazy.

`continue` is also absolutely not the thing you want to do here. that continue just skips the edge in the path which is totally wrong. it means you'll return whatever part of the route is retrievable from cache. the route could be skipping all over the map! if this were to happen you should just:

```c++
throw std::runtime_error("tile cache is connected to tile source via a piece of wet spaghetti and is randomly able and unable to load tiles. please ask the implementer to think a bit more clearly about how he designs a cache");
```

the crux of the issue is that you cant distinguish between a tile you cant get and a tile that doesnt exist

Reply:
Hi @kevinkreiser 

>  in that case shouldnt GetGraphTile block until it can go get the damned tile again

Do you mean that we should prevent any cache evictions until route building is finished?

Reply:
@SiarheiFedartsou that is the assumption that the code has been working on since it was written. if you can do that, that would be ideal because it removes the need for all of these additional checks.

if you absolutely must purge cache while the computation is running then i find it unreasonable that the cache would have been able to get a tile before but cant get it back into cache now. in practical cases most caches will use disk as a place to put the tile when its not in memory. it seems extremely unlikely that you wouldnt be able to read the tile back off of the disk if some other thread cleared it from memory

Reply:
but i think at the end of the day, this is the important part:

> the crux of the issue is that you cant distinguish between a tile you cant get and a tile that doesnt exist

up until now the interface of the graphreader was such that `nullptr` means that the tile DOES NOT EXIST

it seems to me that your use-case is now opening up another possibility of the tile CANNOT BE LOADED

if your tilecache knows that its the latter, then it should throw right away and let that bubble all the way up to the algorithm where the algorithm will be stopped.

basically we change graphreaders interface to be:

nullptr == tile does not exist in this tile set
exception == tile cache failure to get tile

and this has the added benefit of keep the code the way it is and just adding throwing to the tile cache

Reply:
> it seems extremely unlikely that you wouldnt be able to read the tile back off of the disk if some other thread cleared it from memory

Yeah, but it is possible that we could remove it from disk completely, i.e. we will have no chance to get it back(since we will have to load it from the network)

Reply:
sure, but my point is if that happens and you cant get it back from network then you should throw in the tile cache. the problem isnt one you can gracefully handle. if you lose the ability to get tiles you need to just stop doing routing computations because the results of those will be wrong anyway.

Reply:
In this case you are right, and we should throw exception, but we can't do it from cache because we want to build route even some tile doesn't unavailable.